File: backend/src/knowledge/vectorDb.ts
import { ChromaClient, Collection } from 'chromadb';

const COLLECTION_NAME = 'plankton_pv_regulations';

let _client: ChromaClient | null = null;
let _collection: Collection | null = null;

function getClient(): ChromaClient {
  if (!_client) {
    _client = new ChromaClient({ path: process.env.CHROMA_URL ?? 'http://localhost:8000' });
  }
  return _client;
}

export async function getCollection(): Promise<Collection> {
  if (_collection) return _collection;
  const client = getClient();
  _collection = await client.getOrCreateCollection({
    name: COLLECTION_NAME,
    metadata: { 'hnsw:space': 'cosine' },
  });
  return _collection;
}

export async function upsertChunks(
  chunks: {
    id: string;
    text: string;
    metadata: Record<string, string | string[] | null>;
  }[]
): Promise<void> {
  const collection = await getCollection();
  await collection.upsert({
    ids: chunks.map((c) => c.id),
    documents: chunks.map((c) => c.text),
    metadatas: chunks.map((c) => c.metadata as Record<string, string>),
  });
}

export async function deleteCollection(): Promise<void> {
  const client = getClient();
  try {
    await client.deleteCollection({ name: COLLECTION_NAME });
    _collection = null;
  } catch {
    // Collection may not exist yet
  }
}

File: backend/app/services/knowledge/vector_db.py
"""ChromaDB client singleton — port of knowledge/vectorDb.ts."""
from __future__ import annotations

import os
from typing import Any, Optional

import chromadb

COLLECTION_NAME = "plankton_pv_regulations"

_client: Optional[chromadb.AsyncHttpClient] = None
_collection: Optional[Any] = None


def _get_client() -> chromadb.AsyncHttpClient:
    global _client
    if _client is None:
        chroma_url = os.getenv("CHROMA_URL", "http://localhost:8000")
        _client = chromadb.AsyncHttpClient(host=chroma_url.replace("http://", "").split(":")[0],
                                           port=int(chroma_url.split(":")[-1]))
    return _client


async def get_collection() -> Any:
    global _collection
    if _collection is not None:
        return _collection
    client = _get_client()
    _collection = await client.get_or_create_collection(
        name=COLLECTION_NAME,
        metadata={"hnsw:space": "cosine"},
    )
    return _collection


async def upsert_chunks(chunks: list[dict]) -> None:
    """Upsert a list of {id, text, metadata} dicts into ChromaDB."""
    collection = await get_collection()
    await collection.upsert(
        ids=[c["id"] for c in chunks],
        documents=[c["text"] for c in chunks],
        metadatas=[c["metadata"] for c in chunks],
    )


async def delete_collection() -> None:
    global _collection
    client = _get_client()
    try:
        await client.delete_collection(name=COLLECTION_NAME)
        _collection = None
    except Exception:
        pass  # Collection may not exist yet
